### PR TITLE
[JENKINS-27474] Adding build step File Exists

### PR DESCRIPTION
**Pull request by nexus49**
This step verifies if a file exists in the workspace.

https://issues.jenkins-ci.org/browse/JENKINS-27474

### CODE REVIEW

File: basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/FileExistsStep.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2014 Jesse Glick.
```
Comment (jglick) [inline basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/FileExistsStep.java]: Probably you want to add your own copyright. :-)

**Comment by jglick** [comment]
Would need a test. Probably could just be added to `ReadWriteFileStepTest` as a new test case.

**Comment by jglick** [comment]
You forgot to make a `config.jelly`. (`StepConfigTester` can be used to test the correctness of the config form, though I generally do not bother with it for steps with such trivial configuration.)

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ReadWriteFileStepTest.java [hunk 2]
```diff
@@ -36,8 +37,8 @@
     @Rule public JenkinsRule r = new JenkinsRule();
 
     @Test public void basics() throws Exception {
-        WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
-        boolean win = Functions.isWindows();
+        final WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+        final boolean win = Functions.isWindows();
```
Comment (jglick) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ReadWriteFileStepTest.java]: Gratuitous, could be reverted.

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ReadWriteFileStepTest.java [hunk 3]
```diff
@@ -49,4 +50,33 @@
...
+	{
+        final WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition(
+                "node {\n" +
+                "  def exists = fileExists('test.txt')  \n" +
+                "  echo \"FileExists: ${exists}\" \n" +
+                "}"));
+
+		r.assertLogContains("FileExists: false", p.scheduleBuild2(0).get());
+		r.assertBuildStatusSuccess(p.scheduleBuild2(0));
```
Comment (jglick) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ReadWriteFileStepTest.java]: This is building the job twice. Probably you meant something like

```java
r.assertLogContains("FileExists: false", r.assertBuildStatusSuccess(p.scheduleBuild2(0)));
```

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ReadWriteFileStepTest.java [hunk 4]
```diff
@@ -49,4 +50,33 @@
...
+		final WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+		p.setDefinition(new CpsFlowDefinition(
+				"node {\n" +
+				"  writeFile file: 'test2.txt', text:'content of file' \n" +
+				"  def exists = fileExists('test2.txt')  \n" +
+				"  echo \"FileExists: ${exists}\" \n" +
+				"}"));
+
+		r.assertLogContains("FileExists: true", p.scheduleBuild2(0).get());
+		r.assertBuildStatusSuccess(p.scheduleBuild2(0));
```
Comment (jglick) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ReadWriteFileStepTest.java]: (ditto here)

File: aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ReadWriteFileStepTest.java [hunk 5]
```diff
@@ -49,4 +50,33 @@
...
+
+	@Test
+	public void shouldTestFileExistsAndFindFile() throws Exception
+	{
+		final WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+		p.setDefinition(new CpsFlowDefinition(
+				"node {\n" +
+				"  writeFile file: 'test2.txt', text:'content of file' \n" +
+				"  def exists = fileExists('test2.txt')  \n" +
+				"  echo \"FileExists: ${exists}\" \n" +
```
Comment (jglick) [inline aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/ReadWriteFileStepTest.java]: Could save a few seconds of test time and shorten sources by merging these two test cases into one: `fileExists`, `echo`, `writeFile`, `fileExists`, `echo` should print false then true.

File: basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/FileExistsStep.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+    }
+
+    public static final class Execution extends AbstractSynchronousStepExecution<Boolean> {
+
+        @Inject private transient FileExistsStep step;
+        @StepContextParameter private transient FilePath workspace;
+
+        @Override protected Boolean run() throws Exception {
+        	return Boolean.valueOf(workspace.child(step.file).exists());
```
Comment (jglick) [inline basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/FileExistsStep.java]: Can skip `Boolean.valueOf` and rely on autoboxing.

**Comment by jglick** [comment]
I would also merge with `master` to pick up some test fixes.

**Comment by nexus49** [comment]
Thanks Jesse for you feedback. Its my first contribution so i still figure some things out... Let me know if you have further feedback for me to incorporate.

**Pull request merged**